src/editor/Core/LineNumber.re

```diff
@@ -4,10 +4,11 @@
  * Settings and utilities for managing Linumbers
  */
 
+[@deriving (yojson, show)]
 type setting =
-  | On
-  | Off
-  | Relative;
+  | [@name "on"] On
+  | [@name "off"] Off
+  | [@name "relative"] Relative;
```

Comment:
On a similar vein to my previous comment, what are we going to do about "missing" config options relative to VSCode? Ie, VSCode has an `Interval` option. Are we going to strive to have parity? Should we define all options, even if some are a no-op?

Reply:
I think similarity where it makes sense is one thing but at the end of the day this is a different editor with a smaller team and a similar but different goal for my 2 cents copying them to the letter doesn't make sense to me :man_shrugging:

Reply:
It would be amazing if we could support the full breadth of VSCode options (eventually!) But ya, it will be an incremental thing for now - I'm okay with supporting a subset and slowly adding them. We should just try and get as much parity as we can (like using the same options / names as VSCode for these initial set of options). We'll also likely have our own vim-specific (like the tabline mode).

@Akin909 - what's the behavior if a user specifies 'Interval' in the config? Does it crash? Or does it use a different default?

Reply:
@bryphe I've tweaked it to print the error but return a default if the user enters the wrong value, but the configuration that is currently being read is intended to be the default so should be correct

I was separately going to look into editing the default and having it save as your own which would mean changing this to be more robust for incorrect user input situations. I would like to do it in a way that uses the tools `yojson` without writing a lot of per case logic.  Such as expecting a string and checking that and generating a Variant which seems like something that wouldn't scale well, might be the only option but will look into it a bit more as part of that.

Reply:
Great! Thanks for investigating this, @Akin909 !

> I would like to do it in a way that uses the tools yojson without writing a lot of per case logic. Such as expecting a string and checking that and generating a Variant which seems like something that wouldn't scale well, might be the only option but will look into it a bit more as part of that.

This sounds like a good next step. I'm wondering if there are any tools in `yojson` or `ppx_deriving_yojson` that can help keep us from writing a lot of boilerplate for every configuration option. 🤔 

I think this also touches on something @CrossR mentioned - in Oni, we started to transition to specifying configuration values separately instead of in one giant store. Perhaps we could investigate a similar approach here? 

We started an API like `Configuration.registerSetting("editor.lineNumbers", configurationOptions)` - it might be worth exploring again? We might be able to bake in some of that parsing in that layer too. I think this is more similar to how VSCode does it (for example - custom parsing or custom UI for modifying it could live in the `configurationOptions`). Just something to think about.

Reply:
@bryphe that's an interesting idea I was actually reading about lenses in ocaml as a way of creating extensible records :thinking: wonder if that pattern could help in constructing it from multiple different pieces